README.txt
repoze.squeeze
==============

This package provides a WSGI middleware component which "squeezes"
HTML documents by merging browser resources (javascript
and stylesheets).

It uses statistical analysis to determine the optimal bundles based on
the HTML documents that pass through it. Vary-headers are observed, as
are resource expiration dates.

Documents that are not squeezed are given the 'no-cache' pragma in an
expectation that we will be able to squeeze it after sufficient
burn-in. Squeezed documents are served with expiration dates no later
than the expiration dates of the squeezed resources which it
references.

Usage
-----

Configure the middleware with the following two options (required):

  @cache_dir: Relative path to a directory where we store squeezed
  resources

  @url_prefix: Path segment or full base URL that will be used to
  serve the cache directory as static files.

The following option is optional:
  
  @threshold: Floating-point parameter that controls the merge to
  apperances threshold. This parameter can normally be left at the
  default (0.5).

In addition, you must configure the WSGI application to serve the file
cache directory from the ``url_prefix`` (e.g. "cache").

Invalidation
------------

To invalidate the file cache, upstream applications can set the
"X-Squeeze-Invalidate" header to a true value; note that before the
application is called, the middleware adds this key to the environment
such that applications which can't modify the environment can still
flag for invalidation:

  >>> invalidate = environ["X-Squeeze-Invalidate"]
  >>> invalidate()

This will reset the file cache registry; note that no cache files are
ever deleted (to preserve web server integrity). You can manually
purge the file cache at any time be deleting the files (this is
allowed at run-time).


CHANGES.txt
Changelog
---------

0.4.4 (released 2009-06-22)
~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed edge-case where a response would not have a character set
  defined and hence couldn't be expressed as unicode. [malthe]

- Fixed edge-case with self-closing style elements. [malthe]

- Fixed edge-case issue with missing ttl-value. [malthe]

0.4.3 (released 2009-05-11)
~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added environment variable control of file cache
  invalidation. [malthe]

0.4.2 (released 2009-05-11)
~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Add semicolon after javascripts to make sure any function-block is
  properly terminated (necessary due to concatenation). [malthe]

0.4.1 (released 2009-04-20)
~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Parse unicode body to ensure correct character decoding. [malthe]

0.4 (released 2009-04-15)
~~~~~~~~~~~~~~~~~~~~~~~~~

- If a response has status '304 Not Modified', update the ``ttl`` but
  not the content. [malthe]

- If a script- or style-tag does not have a source or link, ignore
  it. [malthe]

- Unquote request URL when querying cache. [malthe]

- Added support for ``repoze.xmliter`` which allows reusing an
  ``lxml`` document tree between middlewares. [malthe]

- Media types for stylesheets are now taken into account when merging
  the resources. Squeze uses CSS2 `@media <type>` syntax. [jvloothuis]

0.3 (released 2008-10-08)
~~~~~~~~~~~~~~~~~~~~~~~~~

- Relative URLs appearing in stylesheets are now made absolute before
  merging. [malthe]

0.2 (released 2008-10-03)
~~~~~~~~~~~~~~~~~~~~~~~~~

- Add setuptools as a dependency. [malthe]

0.1 (released 2008-10-02)
~~~~~~~~~~~~~~~~~~~~~~~~~

- Initial release.




repoze/squeeze/tests.py
import os
import sys
import unittest
import doctest

OPTIONFLAGS = (doctest.ELLIPSIS |
               doctest.NORMALIZE_WHITESPACE)

from tempfile import gettempdir
cache_dir = gettempdir()

class TestResourceSqueezinMiddleware(unittest.TestCase):

    @property
    def middleware(self):
        from repoze.squeeze.processor import ResourceSqueezingMiddleware
        return ResourceSqueezingMiddleware

    @property
    def accept_request_data(self):
        from repoze.squeeze.processor import AcceptRequestData
        return AcceptRequestData

    def test_ie_conditionals(self):
        # IE conditional stylesheets should be left untouched since they
        # shouln't be combined with the other stylesheets.
        html = '''
  <html><head>
    <link type="text/css" media="screen" href="http://john/doe.css">
    <link type="text/css" media="screen" href="http://john/doegh.css">

    <!-- Internet Explorer CSS Fixes -->
    <!--[if lt IE 7]>
        <style type="text/css" media="all">
           @import url(http://foo/bar.css);
    </style>
    <![endif]-->
  </head><body></body></html>'''

        requestdata = self.accept_request_data()

        mw = self.middleware(None, url_prefix='', cache_dir=cache_dir)
        changed, expires, processed_html = mw.process_html(
            requestdata, 'foo', 'http://foo/bar.css', html)
        # It should not handle the IE comment in any way
        self.assertEqual(requestdata.stylesheets.keys(),
                         [('http://john/doe.css', 'http://john/doegh.css')])


if sys.version_info[:3] < (2,5,0):
    print "Python 2.5 is required to run the test suite."
    sys.exit(1)

def test_suite():
    return unittest.TestSuite([
        doctest.DocFileSuite(
        'README.txt',
        globs=dict(cache_dir=cache_dir, os=os),
        optionflags=OPTIONFLAGS,
        package="repoze.squeeze"),

        doctest.DocTestSuite(
        'repoze.squeeze.processor',
        optionflags=OPTIONFLAGS),
        
        unittest.makeSuite(TestResourceSqueezinMiddleware)])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


repoze/squeeze/processor.py
""" Middleware that squeezes resources."""

import re
import os
import urlparse
import lxml.html
import sha
import mimetypes
import logging
import webob
import urllib

logger = logging.getLogger("squeeze")
    
try:
    from repoze.xmliter import XMLSerializer
except ImportError:
    def XMLSerializer(tree, serializer=None):
        return lxml.html.tostring(tree, pretty_print=True)

from cStringIO import StringIO

re_stylesheet_import = re.compile(
    r'^((?:<!--)?\s*@import\surl*\()([^\)]+)(\);?\s*(?:-->)?)$')

re_stylesheet_url = re.compile(
    r'url\((?![A-Za-z]+://)([^\)]+)\)')

class SqueezeInvalidation(object):
    __slots__ = 'invalidate',

    def __init__(self, string_value=None):
        if string_value:
            self.invalidate = string_value.lower() in (
                'invalidate', 'y', 'yes', 't', 'true', 'on', '1')
        else:
            self.invalidate = False

    def __nonzero__(self):
        return self.invalidate

    def __call__(self):
        self.invalidate = True
        
class AcceptRequestData(object):
    def __init__(self):
        self.appearances = {}
        self.javascripts = {}
        self.stylesheets = {}
        self.cache = {}
        
class ResourceSqueezingMiddleware(object):
    invalidation_environ_key = "X-Squeeze-Invalidate"
    
    def __init__(self, app, cache_dir=None, url_prefix=None, threshold=0.5):
        if url_prefix is None:
            raise ValueError("Must configure URL prefix (`url_prefix`).")
        if cache_dir is None:
            raise ValueError("Must configure cache directory (`cache_dir`).")
        
        self.app = app
        self.cache_dir = cache_dir
        self.url_prefix = url_prefix
        self.threshold = threshold
        self.accept_request_registry = {}

    def get_merged_resource(self, cache, selection, mediatypes):
        expires = None
        
        # verify that the entire selection is in the cache
        for url in selection:
            if url not in cache:
                return expires, None

        # create extension
        ext = None
        body, mimetype, ttl = cache[selection[0]]
        if mimetype is not None:
            ext = mimetypes.guess_extension(mimetype)

        # compute digest
        out = StringIO()
        for url, body, mimetype, ttl in ((s,) + cache.get(s)
                                         for s in selection):
            mediatype = mediatypes.get(url)
            if expires is None or ttl < expires:
                expires = ttl
            if mimetype == 'text/css' and mediatype:
                out.write('@media %s {\n%s}\n' % (mediatype, body))
            else:
                out.write(body+'\n')
            if mimetype == 'application/x-javascript':
                out.write(';\n')
        body = out.getvalue()
        digest = sha.new(body).hexdigest() + (ext or "")

        # write file to cache
        filename = os.path.join(self.cache_dir, digest)
        if not os.path.exists(filename):
            file(filename, 'w').write(body)

        return expires, digest

    def get_url_for_resource(self, host, resource):
        if self.url_prefix.startswith('http'):
            return "/".join((self.url_prefix.rstrip('/'), resource))                
        return "/".join((host, self.url_prefix.strip('/'), resource))

    def __call__(self, environ, start_response):
        environ[self.invalidation_environ_key] = SqueezeInvalidation()

        request = webob.Request(environ)
        response = request.get_response(self.app, catch_exc_info=True)

        # get accept request data object based on the value of the vary-header
        vary = map(request.environ.get, response.vary or ('*',))
        accept_request_data = self.accept_request_registry.setdefault(
            ",".join(filter(None, vary)), AcceptRequestData())

        content_type = response.content_type
        if content_type and content_type.startswith('text/html'):
            # check if we need to invalidate cache
            invalidation = environ.get(self.invalidation_environ_key)
            if isinstance(invalidation, basestring):
                invalidation = SqueezeInvalidation(invalidation)

            if invalidation:
                for registry in self.accept_request_registry.values():
                    registry.cache.clear()

            # process document body
            if response.charset is None:
                return
            changed, expires, body = self.process_html(
                accept_request_data, request.host_url,
                request.path, response.unicode_body)

            # if document is unchanged, set the cache-headers as
            if not changed:
                response.cache_control = 'no-cache'

            # set the response expiration date to the minimum value in
            # the set of the response expiration date and the resource
            # expiration date(s).
            if expires and response.expires:
                response.expires = min(response.expires, expires)

            # set new body
            if isinstance(body, unicode):
                response.unicode_body = body
            else:
                response.body = body

        # if url matches a URL we've seen in a processed document, and
        # if it's served from this host, process the response body and
        # cache it
        url = urllib.unquote(request.url)
        if url in accept_request_data.appearances:
            ttl = response.expires
            if content_type == 'text/css':
                base_path = os.path.dirname(url)
                response.body = re_stylesheet_url.sub(
                    'url(%s/\\1)' % base_path, response.body)

            status_code = int(response.status.split(' ', 1)[0])
            if status_code == 200:
                accept_request_data.cache[url] = response.body, content_type, ttl
            elif status_code == 304:
                # update ttl for cached resource
                cache = accept_request_data.cache.get(url)
                if cache is not None:
                    body, content_type, previous_ttl = cache
                    accept_request_data.cache[url] = body, content_type, ttl

        return response(environ, start_response)

    def process_html(self, accept_request_data, host, uri, body):
        javascripts = accept_request_data.javascripts
        stylesheets = accept_request_data.stylesheets
        appearances = accept_request_data.appearances
        cache = accept_request_data.cache

        if not body.strip():
            return False, None, body

        tree = lxml.html.fromstring(body)
        changed = False

        expires = None
        # process javascripts and stylesheets in two loops
        for groups, xpath in (
            (javascripts, './/head/script[@src]'),
            (stylesheets, './/head/link[@href] | .//head/style')):
            items = []
            elements = tree.xpath(xpath)
            for element in elements:
                mutator, accessor = tag_functions[element.tag]
                url = accessor(element)
                if url is not None:
                    items.append(get_url(tree, host, uri, url))
            selections = maintain_appearances(
                items, groups, appearances, self.threshold)
            ttl, changed = self.update_elements(
                elements, selections, tree, host, uri, cache) or changed

            if expires is None or ttl is None or ttl < expires:
                expires = ttl
                
        return changed, expires, XMLSerializer(tree, lxml.html.tostring)

    def update_elements(self, elements, selections, tree, host, uri, cache):
        changed = False
        expires = None
        mediatypes = {}

        for element in elements:
            parent = element.getparent()
            mutator, accessor = tag_functions[element.tag]
            url = accessor(element)
            if url is None:
                continue
            
            # prepend base path to relative path
            src = get_url(tree, host, uri, url)
            mediatypes[src] = element.attrib.get('media')

            if not src.startswith(host):
                continue

            for selection in selections:
                ttl, resource = self.get_merged_resource(cache, selection,
                                                         mediatypes)

                if resource is None or src not in selection:
                    continue

                if expires is None or ttl is None or ttl < expires:
                    expires = ttl

                # if this is the last item in the selection,
                # merge and update link, else remove it.
                if src == selection[-1]:
                    url = self.get_url_for_resource(host, resource)
                    changed = True
                    mutator(element, url)
                else:
                    parent.remove(element)
                    break
                    
        return expires, changed

def update_script_tag(element, url):
    element.attrib['src'] = url

def update_link_tag(element, url):
    element.attrib['href'] = url

def update_style_tag(element, url):
    """Match style tag with regular expression.

    >>> re_stylesheet_import.match('<!-- @import url(foo.css); -->').groups()
    ('<!-- @import url(', 'foo.css', '); -->')
    
    >>> re_stylesheet_import.match('@import url(foo.css);').groups()
    ('@import url(', 'foo.css', ');')
    
    """
    
    m = re_stylesheet_import.match(element.text or u"")
    if m is not None:
        before = m.group(1)
        path = m.group(2)
        after = m.group(3)
        element.text = before+url+after
        
def get_script_url(element):
    return element.attrib['src']

def get_link_url(element):
    return element.attrib['href']

def get_style_url(element):
    m = re_stylesheet_import.match(element.text or u"")
    if m is not None:
        return m.group(2)
    
tag_functions = dict(
    script=(update_script_tag, get_script_url),
    link=(update_link_tag, get_link_url),
    style=(update_style_tag, get_style_url))

def make_squeeze_middleware(app, global_conf, **kw):
    return ResourceSqueezingMiddleware(app, **kw)

def get_url(tree, host, uri, src):
    if src.startswith('http://'):
        return src
    
    if tree.xpath('.//base'):
        url = tree.xpath('.//base')[0].attrib['href']
    else:
        path = "/".join((host, uri))
        protocol, host, path, query, fragment = urlparse.urlsplit(path)
        path = os.path.dirname(path)
        url = urlparse.urlunsplit((protocol, host, path, query, fragment))

    return "/".join((url.rstrip('/'), src))

def get_slices_ordered_by_size(items):
    slices = []
    length = len(items)
    for i in range(length):
        for j in range(i+1, length+1):
            slices.append(tuple(items[i:j]))
    slices.sort(key=lambda item: len(item))
    return slices

def maintain_appearances(refs, groups, appearances, threshold):
    """Update group statistics.

    >>> abcde = tuple('abcde')
    >>> abcd = tuple('abcd')
    >>> abc = tuple('abc')
    >>> de = tuple('de')

    >>> appearances = {}
    >>> groups = {}
    
    >>> maintain_appearances(abcde, groups, appearances, 0.8)
    [('a', 'b', 'c', 'd', 'e')]

    The groups should be updated accordingly.

    >>> groups, appearances
    ({('a', 'b', 'c', 'd', 'e'): 1},
     {'a': 1, 'c': 1, 'b': 1, 'e': 1, 'd': 1})

    Running the same set will update all numbers.
    
    >>> maintain_appearances(abcde, groups, appearances, 0.8)
    [('a', 'b', 'c', 'd', 'e')]

    >>> groups, appearances
    ({('a', 'b', 'c', 'd', 'e'): 2},
     {'a': 2, 'c': 2, 'b': 2, 'e': 2, 'd': 2})

    Now let's run a subset.

    >>> maintain_appearances(abc, groups, appearances, 0.8)
    []

    We now expect to see the inverse group, too.

    >>> groups.keys()
    [('a', 'b', 'c', 'd', 'e'), ('a', 'b', 'c'), ('d', 'e')]
    
    Let's adjust the threshold to see the inclusion of this group.

    >>> maintain_appearances(abc, groups, appearances, 0.4)
    [('a', 'b', 'c')]

    Now, for a real-world scenario.

    >>> for i in range(1000):
    ...     _ = maintain_appearances(abcde, groups, appearances, 0.8)

    >>> for i in range(1000):
    ...     _ = maintain_appearances(abc, groups, appearances, 0.8)

    >>> maintain_appearances(abcde, groups, appearances, 0.8)
    [('a', 'b', 'c'), ('d', 'e')]
    
    >>> appearances
    {'a': 2005, 'c': 2005, 'b': 2005, 'e': 1003, 'd': 1003}

    >>> groups
    {('a', 'b', 'c', 'd', 'e'): 1003,
     ('a', 'b', 'c'): 2003,
     ('d', 'e'): 1001}

    >>> maintain_appearances(abcd, groups, appearances, 0.8)
    [('a', 'b', 'c')]

    >>> maintain_appearances(de, groups, appearances, 0.8)
    [('d', 'e')]
    """

    # increase our appearance count (how many times this
    # resource has passed through the middleware
    for ref in refs:
        appearances.setdefault(ref, 0)
        appearances[ref] += 1

    groups.setdefault(tuple(refs), 0)
            
    selections = []
    # pick out combinations iterator-style
    for selection in get_slices_ordered_by_size(refs):
        if len(selection) == 1:
            continue
        
        value = groups.get(selection)
        if value is not None:
            # update values
            value = groups[selection] = 1 + value

            # calculate "seen together" ratio
            highest = max(map(appearances.get, selection)) or 1

            ratio = float(value)/highest

            if ratio > threshold:
                selections.append(selection)

            for group in groups.keys():
                if group != selection:
                    inverse = tuple(ref for ref in group if ref not in selection)
                    if len(inverse):
                        groups.setdefault(inverse, 0)

    refs = list(refs)
    selections.sort(key=lambda selection: refs.index(selection[0]))
    return selections
